editing/empty-elements-insertion.html

```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE HTML>
+<meta charset=utf-8>
+<title>Placing selection and typing inside empty elements</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+<script src="/resources/testdriver-actions.js"></script>
+<style>
+  #styled {
+    display: inline-block;
+    border: 2px red solid;
+    padding: 1em;
+  }
+</style>
+<div contenteditable>
+  <p><strong id="styled"></strong></p>
+</div>
+<div contenteditable>
+  <p>Lorem <strong id="unstyled"></strong> ipsum</p>
+</div>
+<script>
+  promise_test( async () => {
+    const target = document.querySelector( '#styled' );
+    const actions = new test_driver.Actions();
+    await actions
+      .pointerMove( 0, 0, { origin: target } )
+      .pointerDown( { button: actions.ButtonType.LEFT } )
+      .pointerUp( { button: actions.ButtonType.LEFT } )
+      .send();
+    document.execCommand( 'insertText', false, 'a' );
+    assert_greater_than( target.innerHTML.length, 0, 'The text was inserted into the styled <strong> element' );
+  }, 'Insert text into the inline element styled with border and padding' );
```

Comment:
I think that you should separate test for `<strong>` element which only has `border` or `padding`.

On the other hand, you probably intended to check whether it should or should not work with visible vs. implicitly invisible elements. If so, there could be more tricky cases like `::before` etc?

And I wonder, this is not directly related to editing. If this test requests some browsers to change the behavior, it seems that they need to change the `Selection` result when it's clicked. Perhaps, you should add such test to the directory of Selection API.

Reply:
> I didn't move these tests to the selection API dir, though. As can be seen in a [simple demo](https://jsfiddle.net/Comandeer/z4cs8wm6/), the selection seems to be correct (at least in the case of the visible `span`). In Safari and Chrome this case works, Firefox jumps out of the element when anything is typed inside.

Thank you for explaining the symptom. Indeed, it should work in Firefox too. I'll take a look after current big work ends.

> > If so, there could be more tricky cases like `::before` etc?
> 
> Probably but I don't think such use cases are really sensible in the editing context. [The `::before` pseudoelement is not selectable](https://jsfiddle.net/Comandeer/mr3odq4g/) and I can't think of any real use case for changing that behavior.

If Chrome nor Safari does not treat the elements having only CSS generated content, it makes the fix harder in Firefox because Firefox needs to check whether `::before` and/or `::after` is the only content in it. Therefore, anyway the behavior is important for me to making it fix in Firefox. Therefore, could you add the cases too?

Reply:
Sorry for the delayed response.

I think that I initially misunderstood the case with `::before` and `::after` pseudoelements. For some reason, I thought about putting selection in the pseudoelements. But I suppose you meant being able to put the selection in the empty elements with only the pseudoelements – is my understanding correct? If the latter case is the one you thought of, then I agree that it would make sense to support inserting text also in such empty inline elements.

I added some tests for the second case.

Reply:
> Sorry for the delayed response.

Me too, no worry.

> But I suppose you meant being able to put the selection in the empty elements with only the pseudoelements – is my understanding correct?

Yes, that's right. My intention was adding the case which may create a "frame" for empty element only with CSS styling.

> I added some tests for the second case.

Thank you!